### PR TITLE
Moves the data tree commands, messages, APIs, and tests from salesfor…

### PR DESCRIPTION
**Pull request by shetzel**
…ce-alm to @salesforce/plugin-data

### What does this PR do?
Adds `force:data:tree:export` and `force:data:tree:import` commands
Adds messages
Adds API classes that are not ready to be moved to `@salesforce/data`
Adds unit tests

### What issues does this PR fix or reference?
@W-8449687@

### CODE REVIEW

File: packages/plugin-data/src/commands/force/data/tree/export.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import { Messages, Org } from '@salesforce/core';
+import { ExportApi } from '../../../../lib/data/tree/exportApi';
+
+Messages.importMessagesDirectory(__dirname);
+const messages = Messages.loadMessages('@salesforce/plugin-data', 'tree.export');
+
+export default class Export extends SfdxCommand {
+  public static readonly description = messages.getMessage('description');
+  public static readonly examples = messages.getMessage('examples').split(os.EOL);
+  public static readonly requiresProject = false;
```
Comment (WillieRuemmele) [inline packages/plugin-data/src/commands/force/data/tree/export.ts:21]: Is this defaulted to false? I thought most things were

File: packages/plugin-data/src/api/data/tree/exportApi.ts [hunk 2]
```diff
@@ -0,0 +1,607 @@
...
+export class ExportApi {
+  private logger: Logger;
+
+  private objectTypeRegistry: any = {}; // registry for object type data plan descriptor
+  private referenceRegistry: any = {}; // registry of object type-specific id-to-ref mappings
+  private typeRefIndexes: any = {}; // registry for object type-specific ref counters
+
+  private config!: ExportConfig;
+
+  public constructor(private readonly org: Org, private readonly ux: UX) {
```
Comment (WillieRuemmele) [inline packages/plugin-data/src/api/data/tree/exportApi.ts:49]: Why are we passing in org and ux if they're not used in the constructor?

File: packages/plugin-data/src/api/data/tree/exportApi.ts [hunk 3]
```diff
@@ -0,0 +1,607 @@
...
+  public async export(config: ExportConfig): Promise<AnyJson> {
+    this.config = this.validate(config);
+
+    const { outputDir, plan, query } = this.config;
+
+    if (outputDir) {
+      this.setupOutputDirectory(outputDir);
+    }
+
+    let queryResults: any;
```
Comment (WillieRuemmele) [inline packages/plugin-data/src/api/data/tree/exportApi.ts:68]: how did you get this `any` here? I think there's a `QueryResults` type you can import from jsforce or maybe from core

File: packages/plugin-data/src/api/data/tree/exportApi.ts [hunk 4]
```diff
@@ -0,0 +1,607 @@
...
+    }
+    return this.finalApplyRefs(processedRecordList.records);
+  }
+
+  // Register object types and type hierarchy for plan generation
+  private async recordObjectTypes(recordList: any): Promise<any> {
+    const records = recordList.records;
+
+    if (!records.length) {
+      // TODO: should be on the command
```
Comment (WillieRuemmele) [inline packages/plugin-data/src/api/data/tree/exportApi.ts:157]: why don't we do this now?

File: packages/plugin-data/src/api/data/tree/exportApi.ts [hunk 5]
```diff
@@ -0,0 +1,607 @@
...
+
+  private async processRecordList(recordList: any, parentRef?: any): Promise<any> {
+    // holds transformed sobject tree
+    const sobjectTree = { records: [] };
+
+    // visit each record in the list
+    const processRecordsFn = (record: any) => (): Promise<any> => this.processRecords(parentRef, record, sobjectTree);
+    const recordFns = recordList.records.map((record: any) => processRecordsFn(record));
+
+    // TODO: do we really need sequentialExecute?  Can't we just use Promise.all()?
```
Comment (WillieRuemmele) [inline packages/plugin-data/src/api/data/tree/exportApi.ts:203]: can't we? I feel like this is the time to fix all of these TODOs and make the plugin as pristine as possible

File: packages/plugin-data/src/api/data/tree/exportApi.ts [hunk 6]
```diff
@@ -0,0 +1,607 @@
...
+    return Promise.resolve().then(() => {
+      const field = record[key];
+
+      // skip attributes and id.  Data import does not accept records with IDs.
+      if (key === 'attributes' || key === 'Id') {
+        // If this is an attributes section then we need to add an object reference
+        this.saveRecordRef(record, objRefId);
+      } else {
+        return this.loadMetadata(record.attributes.type)
+          .then((metadata) => {
```
Comment (WillieRuemmele) [inline packages/plugin-data/src/api/data/tree/exportApi.ts:264]: Anyway we can refactor this to get rid of the promise chain?

File: packages/plugin-data/src/api/data/tree/exportApi.ts [hunk 7]
```diff
@@ -0,0 +1,607 @@
...
+            return true;
+          }
+        }
+      }
+
+      return false;
+    });
+
+    if (!result) {
+      const metadataName: string = metadata.name;
```
Comment (WillieRuemmele) [inline packages/plugin-data/src/api/data/tree/exportApi.ts:371]: absolute nit: but why have a variable for this? just print `${metadata.name}`

File: packages/plugin-data/src/api/data/tree/exportApi.ts [hunk 8]
```diff
@@ -0,0 +1,607 @@
...
+          return true;
+        }
+      }
+
+      return false;
+    });
+
+    if (!result) {
+      const metadataName: string = metadata.name;
+      throw new SfdxError(`Unable to find relation for ${metadataName}`);
```
Comment (WillieRuemmele) [inline packages/plugin-data/src/api/data/tree/exportApi.ts:415]: same NIT as above

File: packages/plugin-data/src/api/data/tree/exportApi.ts [hunk 9]
```diff
@@ -0,0 +1,607 @@
...
+
+    return dataPlanPart;
+  }
+
+  /**
+   * This method is used as a second pass to establish references that couldn't be determined
+   * in the initial pass done by processRecordList. It looks for relationship fields that
+   * contain an id.
+   */
+  private finalApplyRefs(sobjectTree: any): { records: any } {
```
Comment (WillieRuemmele) [inline packages/plugin-data/src/api/data/tree/exportApi.ts:534]: probably just me but I feel like this method could use a more descriptive name

File: packages/plugin-data/src/api/data/tree/exportApi.ts [hunk 10]
```diff
@@ -0,0 +1,607 @@
...
+      fileName = path.join(outputDir, fileName);
+    }
+
+    if (has(jsonObject, 'records') && isArray(jsonObject.records)) {
+      recordCount = this.countRecords(jsonObject.records);
+    }
+
+    fs.writeFileSync(fileName, JSON.stringify(jsonObject, null, 4));
+
+    // TODO: move this to the command
```
Comment (WillieRuemmele) [inline packages/plugin-data/src/api/data/tree/exportApi.ts:602]: let's do it

File: packages/plugin-data/src/api/data/tree/importApi.ts [hunk 11]
```diff
@@ -0,0 +1,496 @@
...
+  private sobjectUrlMap: Map<string, string>;
+  private schemaValidator: SchemaValidator;
+  private sobjectTypes: any = {}; // hash of sobject { referenceId: type }
+
+  private config!: ImportConfig;
+
+  // A JsonArray but we need to provide a better type to work with it properly.
+  private importPlanConfig: any;
+
+  public constructor(private readonly org: Org) {
```
Comment (WillieRuemmele) [inline packages/plugin-data/src/api/data/tree/importApi.ts:78]: why are we passing in org if we're not using it?

File: packages/plugin-data/src/api/data/tree/importApi.ts [hunk 12]
```diff
@@ -0,0 +1,496 @@
...
+    this.logger = Logger.childFromRoot(this.constructor.name);
+
+    this.sobjectUrlMap = new Map<string, string>();
+    this.schemaValidator = new SchemaValidator(this.logger, importPlanSchemaFile);
+  }
+
+  /**
+   * Inserts given SObject Tree content into given target Org.
+   *
+   * Validation and options fix-up done in execute() instead
```
Comment (WillieRuemmele) [inline packages/plugin-data/src/api/data/tree/importApi.ts:88]: is this referencing the old `execute` methods that were on commands if so it should be `run` now right?

File: packages/plugin-data/src/api/data/tree/importApi.ts [hunk 13]
```diff
@@ -0,0 +1,496 @@
...
+          msg = JSON.parse(err.message);
+          if (msg.hasErrors && msg.results && msg.results.length > 0) {
+            importResults.errors = msg.results;
+          }
+        } catch (e2) {
+          // throw original
+        }
+      }
+
+      throw err;
```
Comment (WillieRuemmele) [inline packages/plugin-data/src/api/data/tree/importApi.ts:179]: should be `SfdxError.wrap(err)`

File: packages/plugin-data/src/api/data/tree/importApi.ts [hunk 14]
```diff
@@ -0,0 +1,496 @@
...
+      } catch (err) {
+        if (err.name === 'ValidationSchemaFieldErrors') {
+          const e = SfdxError.create('@salesforce/plugin-data', 'importApi', 'dataPlanValidationError', [
+            planPath,
+            err.message,
+          ]);
+          e.name = INVALID_DATA_IMPORT_ERR_NAME;
+          throw e;
+        }
+        throw err;
```
Comment (WillieRuemmele) [inline packages/plugin-data/src/api/data/tree/importApi.ts:235]: same SfdxError.wrap(err)

File: packages/plugin-data/src/api/data/tree/importApi.ts [hunk 15]
```diff
@@ -0,0 +1,496 @@
...
+    // replace with reference found in previously saved records
+    if (resolveRefs && refMap) {
+      // find and stash all '@' references
+
+      // Could optionally use String.match(regex) to avoid turning off the rule
+      /* eslint-disable no-cond-assign */
+      while ((match = refRegex.exec(contentStr))) {
+        foundRefs.add(match[1]);
+      }
+      /* eslint-enable no-cond-assign */
```
Comment (WillieRuemmele) [inline packages/plugin-data/src/api/data/tree/importApi.ts:378]: do we need this rule too?

File: packages/plugin-data/src/commands/force/data/tree/export.ts [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+      description: messages.getMessage('prefix'),
+    }),
+    outputdir: flags.directory({
+      char: 'd',
+      description: messages.getMessage('outputdir'),
+    }),
+  };
+
+  // Overrides SfdxCommand.  This is ensured since requiresUsername == true
+  protected org!: Org;
```
Comment (WillieRuemmele) [inline packages/plugin-data/src/commands/force/data/tree/export.ts:44]: do we need to define this again? I thought it was brought in with SfdxCommad

File: packages/plugin-data/src/commands/force/data/tree/import.ts [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+  id: string;
+}
+
+/**
+ * Command that provides data import capability via the SObject Tree Save API.
+ */
+export default class Import extends SfdxCommand {
+  public static readonly description = messages.getMessage('description');
+  public static readonly examples = messages.getMessage('examples').split(os.EOL);
+  public static readonly requiresProject = false;
```
Comment (WillieRuemmele) [inline packages/plugin-data/src/commands/force/data/tree/import.ts:30]: same as other command comment

File: packages/plugin-data/src/commands/force/data/tree/export.ts (hunk 16 above)
Comment (shetzel) [reply packages/plugin-data/src/commands/force/data/tree/export.ts:44]: Unfortunately yes I think.  SfdxCommand.org is optional and to satisfy the TS compiler I had to override and tell it that `this.org` was not possibly undefined.  Let me know if you have another trick to handle this.

File: packages/plugin-data/src/commands/force/data/tree/export.ts (hunk 1 above)
Comment (shetzel) [reply packages/plugin-data/src/commands/force/data/tree/export.ts:21]: You're right - thanks!  I'll change it.

File: packages/plugin-data/src/commands/force/data/tree/import.ts (hunk 17 above)
Comment (shetzel) [reply packages/plugin-data/src/commands/force/data/tree/import.ts:30]: Yep!  thanks!

File: packages/plugin-data/src/api/data/tree/exportApi.ts (hunk 2 above)
Comment (shetzel) [reply packages/plugin-data/src/api/data/tree/exportApi.ts:49]: They are.  this is shorthand TS syntactic sugar that will add them as readyonly instance vars.

File: packages/plugin-data/src/api/data/tree/exportApi.ts (hunk 3 above)
Comment (shetzel) [reply packages/plugin-data/src/api/data/tree/exportApi.ts:68]: Because I didn't want to spend too much time typing everything when there are stories to handle refactoring all this into `@salesforce/data`, and I think this is just internal.  Still useful though so I'll change it.

File: packages/plugin-data/src/api/data/tree/exportApi.ts (hunk 4 above)
Comment (shetzel) [reply packages/plugin-data/src/api/data/tree/exportApi.ts:157]: You're right, and I tried to but it was causing lots of issues so I punted.  There are more work items to handle things like this so unless we really need to I'd prefer to defer.

File: packages/plugin-data/src/api/data/tree/importApi.ts (hunk 13 above)
Comment (shetzel) [reply packages/plugin-data/src/api/data/tree/importApi.ts:179]: Good eye - thanks!

File: packages/plugin-data/src/commands/force/data/tree/export.ts (hunk 16 above)
Comment (amphro) [reply packages/plugin-data/src/commands/force/data/tree/export.ts:44]: This is the trick I have been using too.

File: packages/plugin-data/src/api/data/tree/importApi.ts (hunk 15 above)
Comment (shetzel) [reply packages/plugin-data/src/api/data/tree/importApi.ts:378]: Probably not.  I'll try removing - thanks!

File: packages/plugin-data/src/api/data/tree/exportApi.ts (hunk 7 above)
Comment (shetzel) [reply packages/plugin-data/src/api/data/tree/exportApi.ts:371]: because of linting.  There is a rule about a string template using a non-string.  The type of metadata.name is any, so I can either do `${metadata.name as string}` or the code above (or better typing overall, but that's another effort).

File: packages/plugin-data/src/api/data/tree/exportApi.ts (hunk 10 above)
Comment (shetzel) [reply packages/plugin-data/src/api/data/tree/exportApi.ts:602]: I agree, and I tried but it causes a huge snowball that I thought best to handle in the refactor work items.

File: packages/plugin-data/src/api/data/tree/importApi.ts (hunk 12 above)
Comment (shetzel) [reply packages/plugin-data/src/api/data/tree/importApi.ts:88]: Oh good catch.  I renamed some methods within this class.  There used to be one called `execute`.  This comment should just go away because it refers to very legacy things.

File: packages/plugin-data/package.json [hunk 18]
```diff
@@ -28,7 +28,15 @@
                 "description": "fetch records using SOQL"
               },
               "tree": {
-                "description": "manipulate records using the tree API"
+                "description": "import/export records using the tree API",
+                "subtopics": {
```
Comment (amphro) [inline packages/plugin-data/package.json:32]: Why are these here? They aren't topics right?

File: packages/plugin-data/src/api/data/tree/exportApi.ts (hunk 6 above)
Comment (shetzel) [reply packages/plugin-data/src/api/data/tree/exportApi.ts:264]: Yes but I'd like to save that for the refactor work item since it's not super straight forward.

**Review by amphro — APPROVED** [review]
I basically didn't really review this since I think you are doing as much of a copy/paste over as possible.

File: packages/plugin-data/package.json (hunk 18 above)
Comment (shetzel) [reply packages/plugin-data/package.json:32]: No they aren't topics.  It's a mistake.  I'll remove, thanks!

**Review by WillieRuemmele — APPROVED** [review]
